src/backend/distributed/metadata/node_metadata.c

```diff
@@ -692,10 +694,22 @@ master_update_node(PG_FUNCTION_ARGS)
 		LockShardsInPlacementListMetadata(placementList, AccessExclusiveLock);
 	}
 
+	char *oldWorkerName = pstrdup(workerNode->workerName);
+	int oldWorkerPort = workerNode->workerPort;
+
+	/* make sure we don't have any lingering session lifespan connections */
+	CloseNodeConnectionsAfterTransaction(oldWorkerName, oldWorkerPort);
+
+	/*
+	 * We're not sure that there is no concurrent queries going on, so remove all
+	 * entries in shared connection stats for the current database.
+	 */
+	RemoveAllSharedConnectionEntriesForNode(oldWorkerName, oldWorkerPort);
```

Comment:
I don't think this works. SELECTs will still be opening new connections after this call and TryToIncrementSharedConnectionCounter will happily add back the entry.

Reply:
We seem to lock all  the placements with `AccessExclusiveLock ` above, and `UpdateNodeLocation ` right below?

Reply:
Selects don't take metadata locks. No lock does not conflict with AccessExclusiveLock.

Reply:
Yes, but SELECTs don't take shard metadata locks at all.

Reply:
I'm wondering whether we should add the  entries in a different way such that the absence of an entry when we open a connection means that we should not / no longer count.

Reply:
How can we differentiate that case vs the first time a connection is established to a worker node? If we can find a way to do that, that'd be a simple/nice approach.

One other idea that I derived from the above is to add one more field to the entry `isActive` or such, which we set it to `false` here and `master_remove_node`. Similar to you suggestion, we do not count for `isActive=false` nodes anymore.

And, at the start of `master_add_node/master_remove_node/master_update_node`(anymore UDF that touches node metadata?) , we can go thorough the hash map, and remove `isActive=false` entries. So that we have almost always have a very high utilization of the hashmap.


Does that sound OK?

Reply:
Yes.